runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkExecutableStageFunction.java

```diff
@@ -176,7 +184,23 @@ private StateRequestHandler getStateRequestHandler(
     } catch (IOException e) {
       throw new RuntimeException("Failed to setup state handler", e);
     }
+
+    // Need to discard the old key's state
+    if (bagUserStateHandlerFactory != null) {
+      bagUserStateHandlerFactory.resetForNewKey();
+    }
+    final StateRequestHandler userStateHandler;
+    if (executableStage.getUserStates().size() > 0) {
+      bagUserStateHandlerFactory = new InMemoryBagUserStateFactory();
```

Comment:
I believe this only needs to be done once if you reset the `bagUserStateHandlerFactory` above.

Reply:
If this is this case, we should change it in the Flink runner too, as it looks like a new `InMemoryBagUserStateFactory` is being created each time: https://github.com/apache/beam/blob/9c5f768dfed261a172a9cc29365e128b93a0979a/runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java#L176

Reply:
The above is in the `open()` method which is only called once for the lifetime of the pipeline, not per key.

Reply:
Right, understood.